hal/armv8m/nrf/hal.c

```diff
@@ -0,0 +1,199 @@
+{
+	return addr;
+}
+
+
+void hal_kernelGetEntryPointOffset(addr_t *off, int *indirect)
+{
+	*off = sizeof(unsigned);
+	*indirect = 1;
+}
+
+
+void hal_kernelEntryPoint(addr_t addr)
+{
+	hal_common.entry = addr;
+}
+
+
+int hal_memoryAddMap(addr_t start, addr_t end, u32 attr, u32 mapId)
+{
+	return 0;
+}
+
+
+static void hal_getMinOverlappedRange(addr_t start, addr_t end, mapent_t *entry, mapent_t *minEntry)
+{
+	if ((start < entry->end) && (end > entry->start)) {
+		if (start > entry->start) {
+			entry->start = start;
+		}
+
+		if (end < entry->end) {
+			entry->end = end;
+		}
+
+		if (entry->start < minEntry->start) {
+			minEntry->start = entry->start;
+			minEntry->end = entry->end;
+			minEntry->type = entry->type;
+		}
+	}
+}
+
+
+int hal_memoryGetNextEntry(addr_t start, addr_t end, mapent_t *entry)
+{
+	int i;
+	mapent_t tempEntry, minEntry;
+
+	static const mapent_t entries[] = {
+		{ .start = (addr_t)__init_start, .end = (addr_t)__init_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__text_start, .end = (addr_t)__etext, .type = hal_entryTemp },
+		{ .start = (addr_t)__rodata_start, .end = (addr_t)__rodata_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__init_array_start, .end = (addr_t)__init_array_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__fini_array_start, .end = (addr_t)__fini_array_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__ramtext_start, .end = (addr_t)__ramtext_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__data_start, .end = (addr_t)__data_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__bss_start, .end = (addr_t)__bss_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__heap_base, .end = (addr_t)__heap_limit, .type = hal_entryTemp },
+		{ .start = (addr_t)__stack_limit, .end = (addr_t)__stack_top, .type = hal_entryTemp },
+	};
+
+	if (start == end) {
+		return -1;
+	}
+
+	hal_memset(&tempEntry, 0, sizeof(tempEntry));
+	hal_memset(&minEntry, 0, sizeof(minEntry));
+	minEntry.start = (addr_t)-1;
+
+	/* Syspage entry */
+	tempEntry.start = (addr_t)hal_common.hs;
+	tempEntry.end = (addr_t)__heap_limit;
+	tempEntry.type = hal_entryReserved;
+	hal_getMinOverlappedRange(start, end, &tempEntry, &minEntry);
+
+	for (i = 0; i < sizeof(entries) / sizeof(entries[0]); ++i) {
+		if (entries[i].start >= entries[i].end) {
+			continue;
+		}
+		hal_memcpy(&tempEntry, &entries[i], sizeof(mapent_t));
+		hal_getMinOverlappedRange(start, end, &tempEntry, &minEntry);
+	}
+
+	if (minEntry.start != (addr_t)-1) {
+		entry->start = minEntry.start;
+		entry->end = minEntry.end;
+		entry->type = minEntry.type;
+
+		return 0;
+	}
+
+	return -1;
+}
+
+
+void hal_cpuReboot(void)
+{
+	hal_cpuReset();
+	__builtin_unreachable();
+}
+
+
+/* clang-format off */
+int hal_cpuJump(void)
+{
+	if (hal_common.entry == (addr_t)-1) {
+		return -1;
+	}
+
+	hal_interruptsDisableAll();
+
+	__asm__ volatile("mov r9, %1; \
+		 blx %0"
+		:
+		: "r"(hal_common.entry), "r"((addr_t)hal_common.hs));
```

Comment:
imho r9 should in clobbers

Reply:
I had to think about it and I agree on that it's better to add clobbers with r9 here (based on https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html). I would also change it for other target architectures, so I've done it in the separate commit.

Reply:
For Cortex-M processors, the `r9` register in clobbers is no nedded (the compiler will ignore it because it is run with
`-ffixed-r9` option).
Formally, the register `lr` should be added to the clobbers list ( the `blx` instruction is used in inline assemler ).
If it certain that there will never be return from a function called in `blx` instruction, these registers do not need to be included in the clobbers list.

If the above condition is met, perhaps a better solution would be, for example, this:
```
static void _hal_cpuJump(void)
{
	__asm__ volatile("mov r9, %1; \
		 blx %0"
		:
		: "r"(hal_common.entry), "r"((addr_t)hal_common.hs));

	__builtin_unreachable();
}

int hal_cpuJump(void)
{
	if (hal_common.entry == (addr_t)-1)
		return -1;

	hal_interruptsDisableAll();

	mpu_getHalData(hal_common.hs);

	_hal_cpuJump();

	return 0;
}
```
Instead of `blx` you can use `bx`.

Reply:
@kemonats is right, you don't need to add `r9` to clobbers because of `-ffixed-r9` flag (it prohibits the compiler to allocate this register), I forgot this flag is given to plo too (it's only needed in kernel because of .got pointer being placed in r9), my bad. I also agree that `bx` more appropriate choice.

Reply:
Ok, I've got this. Thanks @kemonats for the suggestion, I think I've used `bx` and `__builtin_unreachable();`, but without separate function.